File: 이상훈/1주차/2회차/정렬리스트병합.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        void AddNode(ListNode **tail, ListNode **l)
        {
                ListNode *newNode = new ListNode((*l)->val);
                (*tail)->next = newNode;
                *tail = newNode;
                *l = (*l)->next;
        }

        ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {

            ListNode *head = new ListNode();
            ListNode *tail = head;

            while (l1 != nullptr && l2 != nullptr) {
                if (l1->val > l2->val) AddNode(&tail, &l2);
                else AddNode(&tail, &l1);
            };

            while (l1 != nullptr) AddNode(&tail, &l1);
            while (l2 != nullptr) AddNode(&tail, &l2);

            return head->next;
        }
};

File: 이상훈/1주차/2회차/정렬리스트병합.py
# https://leetcode.com/problems/merge-two-sorted-lists/submissions/

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def mergeTwoLists( list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    head = ListNode()
    ptr = head

    while list1 and list2:
        if list1.val < list2.val:
            head.next = list1
            list1 = list1.next
        else:
            head.next = list2
            list2 = list2.next

        head = head.next

    if not list1:
        head.next = list2
    else:
        head.next = list1

    return ptr.next
